integrations/cdi/common-cdi/reference-counted-context/src/main/java/io/helidon/integrations/cdi/referencecountedcontext/ReferenceCountedContext.java

```diff
@@ -0,0 +1,474 @@
+/*
+ * Copyright (c) 2019 Oracle and/or its affiliates. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.helidon.integrations.cdi.referencecountedcontext;
+
+import java.lang.annotation.Annotation;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+
+import javax.enterprise.context.spi.AlterableContext;
+import javax.enterprise.context.spi.Contextual;
+import javax.enterprise.context.spi.CreationalContext;
+import javax.enterprise.inject.spi.BeanManager;
+
+/**
+ * A somewhat special-purpose {@link AlterableContext} that
+ * {@linkplain #destroy(Contextual) destroys} a contextual instance
+ * when its thread-specific reference count drops to zero or less than
+ * zero.
+ *
+ * <p>A contextual instance's thread-specific reference count is
+ * incremented when either the {@link #get(Contextual)} or {@link
+ * #get(Contextual, CreationalContext)} method is called.  It is
+ * decremented when the obtained instance is passed to the {@link
+ * #decrementReferenceCount(Contextual)} method, which is indirectly
+ * and solely responsible for that instance's ultimate removal from
+ * this {@link ReferenceCountedContext}.</p>
+ *
+ * <p>Internally, the {@link #destroy(Contextual)} method simply calls
+ * {@link #decrementReferenceCount(Contextual)}.</p>
+ *
+ * <h2>Thread Safety</h2>
+ *
+ * <p>Instances of this class are safe for concurrent use by multiple
+ * threads.</p>
+ *
+ * @see #decrementReferenceCount(Contextual)
+ *
+ * @see #destroy(Contextual)
+ *
+ * @see #get(Contextual, CreationalContext)
+ *
+ * @see ReferenceCounted
+ */
+public final class ReferenceCountedContext implements AlterableContext {
+
+
+    /*
+     * Static fields.
+     */
+
+
+    /**
+     * A {@link ThreadLocal} containing a {@link Map} whose keys are
+     * instances of this class, and whose values are {@link Map}s
+     * indexing {@link Instance}s by {@link Contextual} instances.
+     */
+    private static final ThreadLocal<Map<ReferenceCountedContext, Map<Contextual<?>, Instance<?>>>> ALL_INSTANCES =
```

Comment:
I'm struggling to understand the relationship between `ReferenceCountedContext` instances and thread specific values. My initial reaction to this `ThreadLocal` was to wonder why the `Map<Contextual<?>, Instance<?>>` cannot just be a field of the `ReferenceCountedContext` itself.

I assume that there really is a 1:N relationship, but there are too many indirections going on for my tired brain to see it...

Reply:
Yes.  Somewhere I read and internalized that `ThreadLocal`s should _always_ be `static`—like _always_, even at the expense of a design like this.

If we hold that constant for the moment, then the key to the `Map` held by the `ThreadLocal` is simply a `ReferenceCountedContext` instance.  In normal circumstances, there will only be one entry (in a CDI container there is normally only one instance of a `Context` implementation).

What you are seeing here is the idiom suggested by those championing the use of `static` `ThreadLocal`s even when logically their contents are supposed to be affiliated with a particular instance.

So what we're left with is: a thread-safe association between a `ReferenceCountedContext` (usually just one) and a `Map` that indexes `Instance`s by `Contextual`s (CDI `Bean`s).

You are correct that logically speaking a `Context` instance, should "have" a single `Map` that indexes `Instance`s by `Contextual`s—per thread.

See http://www.0xcafefeed.com/2004/06/of-non-static-threadlocals-and-memory/ for one arbitrarily suggested discussion about `static` and non-`static` `ThreadLocal`s that influenced my design here.

Reply:
Yes, I always make `ThreadLocal` instances `static`. That wasn't my confusion here, but I think you answered the question: a single `ReferenceCountedContext` instance can be shared across threads, but the state must be thread-specific.

Reply:
Exactly right.